### ISSUE TITLE
Mixed modality training data

### ISSUE THREAD

**Issue by haon-chen** [2U1/Llama3.2-Vision-Finetune#18]
This code does not support mixed modality training data (e.g., some data in a batch have images while some don't). Could you modify your code to support batching training of mixed modality data?

**Comment by 2U1** [comment]
I am aware of the issue. I'll try to address it.
Thanks.

**Comment by 2U1** [comment]
I'm not availavble with my server now, So may I ask you that zero2 dose not work with mixed modality training too?

**Comment by haon-chen** [comment]
No, it will raise errors like 
"File "/home/aiscuser/.local/lib/python3.10/site-packages/transformers/models/mllama/modeling_mllama.py", line 595, in forward
attn_output = torch.nn.functional.scaled_dot_product_attention(
RuntimeError: The expanded size of the tensor (3075) must match the existing size (4100) at non-singleton dimension 3.  Target sizes: [12, 32, 48, 3075].  Tensor sizes: [9, 1, 27, 4100]".

**Comment by haon-chen** [comment]
You are right, it works with zero2 after I change my data collator to add dummy tensors for empty images (I think you should add this to your code :) ). Do you have any ideas why the program will crash with zero3?

**Comment by 2U1** [comment]
@haon-chen I thought of adding dummy tensor but, when adding that I think the model code should be fixed for only creating the activation flow for the dummy and not using it. (I'm not sure the model dosen't use the dummy data without <image> token.)

zero3 should has the same activation flow in all datas in the batch. To the best of my knowledge it devides the tensor and also the optimizer states across the gpu. 

If it dosen't harms the models performance than I'll add that. Thanks for letting me know.

**Comment by haon-chen** [comment]
@2U1 FYI, I add dummy tensors by setting image tensors like pixel values to all ones, and cross attention masks to all zeros. I try to make model ignore these dummy tensors. I hope this helps :).

**Comment by xjohnxjohn** [comment]
@haon-chen Do you have sample for "dummy tensors"?

**Comment by haon-chen** [comment]
@xjohnxjohn I simply set pixel_values and aspect_ratio_ids to all ones, aspect_ratio_mask and batch_cross_attention_mask to all zeros.

**Comment by 2U1** [comment]
@haon-chen Sorry for the late reply. It's good idea for making the `cross_attention_mask` to all zeros. 
It can work similar in other VLMs. I'll workaround with it soon.

**Comment by 2U1** [comment]
I've updated the code for supporting mixed-modality data. There were some other issues when making the dataset, so I've fixed it together. 
I think it should work for now.

**Comment by haon-chen** [comment]
@2U1 @xjohnxjohn I have released a mllama-based multimodal embedding model https://huggingface.co/intfloat/mmE5-mllama-11b-instruct. The training codes are in https://github.com/haon-chen/mmE5. Feel free to try them out!

**Comment by 2U1** [comment]
@haon-chen Thanks for the great job!
